ipam/allocator.go

```diff
@@ -159,97 +187,293 @@ const (
+func (a *Allocator) GetDefaultAddressSpaces() (string, string, error) {
+	return localAddressSpace, globalAddressSpace, nil
+}
+
+// RequestPool returns an address pool along with its unique id.
+func (a *Allocator) RequestPool(addressSpace, pool, subPool string, options map[string]string, v6 bool) (string, *net.IPNet, map[string]string, error) {
+	var (
+		nw, pw   *net.IPNet
+		ipr      *addressRange
+		c, p, pp *PoolData
+		ok       bool
+		err      error
+	)
 
-// AddSubnet adds a subnet for the specified address space
-func (a *Allocator) AddSubnet(addrSpace AddressSpace, subnetInfo *SubnetInfo) error {
-	// Sanity check
-	if addrSpace == "" {
-		return ErrInvalidAddressSpace
+	if addressSpace == "" {
+		return "", nil, nil, ipamapi.ErrInvalidAddressSpace
 	}
-	if subnetInfo == nil || subnetInfo.Subnet == nil {
-		return ErrInvalidSubnet
+
+	if pool == "" && subPool != "" {
+		return "", nil, nil, ipamapi.ErrInvalidSubPool
 	}
-	// Convert to smaller internal subnets (if needed)
-	subnetList, err := getInternalSubnets(subnetInfo.Subnet, a.internalHostSize)
-	if err != nil {
-		return err
+
+	if pool != "" {
+		_, nw, err = net.ParseCIDR(pool)
+		if err != nil {
+			return "", nil, nil, ipamapi.ErrInvalidPool
+		}
+		if pw, err = adjustAndCheckSubnetSize(nw); err != nil {
+			return "", nil, nil, err
+		}
+		// Reset pool to canonical form
+		pool = types.GetIPNetCanonical(nw).String()
+		if subPool != "" {
+			_, sub, err := net.ParseCIDR(subPool)
+			if err != nil {
+				return "", nil, nil, ipamapi.ErrInvalidSubPool
+			}
+			lIP, e := types.GetHostPartIP(sub.IP, sub.Mask)
+			if e != nil {
+				return "", nil, nil, fmt.Errorf("failed to compute range's lowest ip address: %v", e)
+			}
+			bIP, e := types.GetBroadcastIP(sub.IP, sub.Mask)
+			if e != nil {
+				return "", nil, nil, fmt.Errorf("failed to compute range's broadcast ip address: %v", e)
+			}
+			hIP, e := types.GetHostPartIP(bIP, sub.Mask)
+			if e != nil {
+				return "", nil, nil, fmt.Errorf("failed to compute range's highest ip address: %v", e)
+			}
+			ipr = &addressRange{ipToUint32(types.GetMinimalIP(lIP)), ipToUint32(types.GetMinimalIP(hIP))}
+		}
+	} else {
+		pool, nw, err = a.getPredefinedPool(addressSpace, v6)
+		if err != nil {
+			return "", nil, nil, err
+		}
+		if pw, err = adjustAndCheckSubnetSize(nw); err != nil {
+			return "", nil, nil, err
+		}
 	}
+
+	k := subnetKey{addressSpace: addressSpace, subnet: pool, childSubnet: subPool}
+
 retry:
-	if a.contains(addrSpace, subnetInfo) {
-		return ErrOverlapSubnet
+	if ipr == nil {
+		a.Lock()
+		if p, ok = a.subnets[k]; ok {
+			p.RefCount++
+			goto finalize
+		}
+		a.Unlock()
+		if a.contains(addressSpace, nw) {
+			return "", nil, nil, ipamapi.ErrPoolOverlap
+		}
 	}
 
-	// Store the configured subnet and sync to datatstore
-	key := subnetKey{addrSpace, subnetInfo.Subnet.String(), ""}
 	a.Lock()
-	a.subnets[key] = subnetInfo
+	if p, ok = a.subnets[k]; ok {
```

Comment:
This part of the code needs rewrite. Too many gotos back and forth make the code completely unreadable and more importantly error prone since you also acquire `a.Lock` before jumping to `finalize` and then you `a.Unlock` which will is just a bug magnet